### ISSUE TITLE
Add interactive mode to the CLI

### ISSUE THREAD

**Issue by mlorbetske** [dotnet/templating#179]
- Use the host specific file to encode the progression of choices
- Add a menu system
- Add parameter specific input gestures & corresponding selection mechanisms to the host files
- Update all templates with supporting data for interactive mode

**Comment by RehanSaeed** [comment]
If you are going to implement this, then you should also consider that some symbols are dependant on others e.g. I have a symbol to enable HTTPS and another to enable a HSTS HTTP header which is only applicable if HTTPS is enabled. Ideally there should be some way of relating the two symbols. I propose the following condition property:

```
"HttpsEverywhere": {
  "type": "parameter",
  "datatype": "bool",
  "defaultValue": "true"
},
"Strict Transport Security (HSTS)": {
  "condition": "HttpsEverywhere && SomethingElse"
  "type": "parameter",
  "datatype": "bool",
  "defaultValue": "true"
}
```

Here the HSTS symbol is only shown to the user if they enabled the HttpsEverywhere feature. Presumably you would show the symbols in the order they appear in `template.json`.

**Comment by mlorbetske** [comment]
These interaction scripts/UI hints (for non-CLI cases) would be specified in the host specific files (see https://github.com/dotnet/templating/blob/master/template_feed/Microsoft.DotNet.Web.ProjectTemplates.1.x/content/EmptyWeb-CSharp/.template.config/dotnetcli.host.json for what host-specific files look like today).

For the actual prompting for interactive mode at the CLI, we're very much keeping in mind prompting order and conditional parameters like you mention with the switch for HSTS. Hopefully over the next week we'll get a chance to pull our current thinking into a proposal and add it to this issue for comment.

@seancpeters

**Comment by jerriep** [comment]
I would like to see this allow code execution as well. So in other words, I do not want to present the user with a static list, but instead on a dynamic list which I obtain from an external system. Let me give you a sample use case:

At [Auth0](https://auth0.com/) we use the standard ASP.NET Core OIDC or JWT middleware (depending on the type of application) to allow users to secure their applications 

I would like to use the templating engine to pre-configure the generated project for the user's particular environment (i.e. the Application they created in Auth0). 

So when the user executes the template, I want to reach out to the Auth0 backend to retrieve the list of Applications a user has configured in Auth0. I then would like the template to display the list of Applications, and allow the user to select the one they want to configure.

Once the user selects from the list, I can then generate the project with the configuration values for the Application they selected pre-configured.

Hope that makes sense :)

**Comment by mlorbetske** [comment]
Hi @jerriep, thanks for the input! We're actively avoiding doing things that allow for the execution of arbitrary code but do have some designs on consuming values from feeds/external files to populate argument values or be used in a future template operation that we're in the process of designing.

For certain templates that we don't yet have in the CLI, we have a similar issue - we need to call an external API to get some stuff set up, retrieve values from that call & supply them to the template so that the retrieved values can be incorporated into the generated content. Our current plan for this is to release the portion that calls the API as a separate tool & have it produce an arguments file that is expanded and passed in as parameters to the template. This approach has multiple benefits:
* If the user creating the template does not have direct access to the API, the file can be generated by a user that does & supplied to the user creating the template
* The user isn't blocked from creating the template in a non-interactive/script type scenario because the interactive prompts that would call the API wouldn't be available
* The user can run the tool again later to change the parameters (setting up a deployment environment perhaps)
* The tool/logic can likely be shared with other consumers of the template & engine, broadening the reach of the tool beyond just the CLI

I'm certainly interested to see if our approach for retrieving values from feeds meets your needs or if we can extend it in some way such that it does. Otherwise, investigating the standalone tool approach as we will be is the preferred option. We'd be happy to help with getting that figured out if you need, especially since we'll (as mentioned) be starting down that road soon ourselves.

**Comment by jerriep** [comment]
@mlorbetske I think generating a file would not be ideal in our scenario. The information which is being retrieved in this case is potentially sensitive (i.e. it may contain client secrets), so having a file lie around on disc containing all this information is not a good thing.

**Comment by Odonno** [comment]
`dotnet new` really (really really) miss an interactive mode like yeoman. How is this feature going on? And hopefully, how can we help?

**Comment by grinrag** [comment]
How does this fit the experienses track for the next releases?

**Comment by grinrag** [comment]
The part of the issue is resolved with the PR #2276

**Comment by vlada-shubina** [comment]
This functionality to be implemented as part of CLI, and not as part of template engine
